rapids-cmake/test/add.cmake

```
@@ -0,0 +1,149 @@
+
+  if(NOT DEFINED _RAPIDS_TEST_NAME)
+    message(FATAL_ERROR "rapids_add_test called without a name")
+  endif()
+
+  if(NOT DEFINED _RAPIDS_TEST_COMMAND)
+    message(FATAL_ERROR "rapids_add_test called without a command")
+  endif()
+
+  list(POP_FRONT _RAPIDS_TEST_COMMAND command_or_target)
+  set(args "${_RAPIDS_TEST_COMMAND}")
+
+  set(command ${command_or_target})
+  if(TARGET ${command_or_target})
+    set(command "$<TARGET_FILE:${command}>")
+  endif()
+
+  if(NOT DEFINED _RAPIDS_TEST_WORKING_DIRECTORY)
+    set(_RAPIDS_TEST_WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}")
+  endif()
+
+  # Provide a copy of the test runner in the binary directory so that tests still can be executed if
+  # for some reason rapids-cmake src has been removed.
+  set(_rapids_run_gpu_test_script "${PROJECT_BINARY_DIR}/rapids-cmake/run_gpu_test.cmake")
+  set(_rapids_run_gpu_test_script_for_install "./run_gpu_test.cmake")
+  if(NOT EXISTS "${_rapids_run_gpu_test_script}")
+    file(COPY "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/detail/run_gpu_test.cmake"
+         DESTINATION "${PROJECT_BINARY_DIR}/rapids-cmake/")
+  endif()
+
+  add_test(NAME ${_RAPIDS_TEST_NAME}
+           COMMAND ${CMAKE_COMMAND} "-Dcommand_to_run=${command}" "-Dcommand_args=${args}" -P
+                   "${_rapids_run_gpu_test_script}"
+           WORKING_DIRECTORY "${_RAPIDS_TEST_WORKING_DIRECTORY}")
+
+  include(${CMAKE_CURRENT_FUNCTION_LIST_DIR}/gpu_requirements.cmake)
+  if(DEFINED _RAPIDS_TEST_GPUS)
+    rapids_test_gpu_requirements(${_RAPIDS_TEST_NAME} GPUS ${_RAPIDS_TEST_GPUS}
+                                 PERCENT ${_RAPIDS_TEST_PERCENT})
+  endif()
+
+  if(_RAPIDS_TEST_INSTALL_COMPONENT_SET)
+    include(${CMAKE_CURRENT_FUNCTION_LIST_DIR}/detail/record_test_component.cmake)
+    include(${CMAKE_CURRENT_FUNCTION_LIST_DIR}/detail/record_test_command.cmake)
+    include(${CMAKE_CURRENT_FUNCTION_LIST_DIR}/detail/record_install.cmake)
+
+    if(NOT TARGET rapids_test_install_${_RAPIDS_TEST_INSTALL_COMPONENT_SET})
+      add_library(rapids_test_install_${_RAPIDS_TEST_INSTALL_COMPONENT_SET} INTERFACE)
+    endif()
+
+    if(TARGET ${command_or_target})
+      get_target_property(output_name ${command_or_target} OUTPUT_NAME)
+      if(output_name)
+        set(command_for_install "./${output_name}")
+      else()
+        set(command_for_install "./${command_or_target}")
+      endif()
+      rapids_test_record_install(TARGET ${command_or_target} COMPONENT
+                                 ${_RAPIDS_TEST_INSTALL_COMPONENT_SET})
+    else()
+      set(command_for_install ${command_or_target})
+    endif()
+    rapids_test_record_test_component(NAME ${_RAPIDS_TEST_NAME} COMPONENT
+                                      ${_RAPIDS_TEST_INSTALL_COMPONENT_SET})
+    rapids_test_record_test_command(NAME
+                                    ${_RAPIDS_TEST_NAME}
+                                    COMMAND
+                                    cmake
```

Better to be `CMAKE_COMMAND`?
```suggestion
                                    ${CMAKE_COMMAND}
```

---

Actually, even better might be that we save the command used for `add_test` and then reuse it here.

---

We have to replicate the logic for mapping from a cmake target to actual executable name that `add_test` does internally. That is why we can't just re-use the same arguments provided to `add_test`.

( Needed as the target name and output executable name don't need to match ).

---

We also can't re-use `CMAKE_COMMAND` since it encodes the absolute path to cmake on the build machine which might not be the same.

We also can't ask `CMAKE_COMMAND` to be evaluated when executing the test as that variable doesn't exist when ctest executes the test file ( No built-in cmake variables exist in this mode )